contrib/hadoop-store-builder/src/java/voldemort/store/readonly/disk/HadoopStoreWriter.java

```diff
@@ -132,56 +130,69 @@ private void initFileStreams(boolean isCompressionEnabled, String compressionCod
                 && compressionCodec.toUpperCase(Locale.ENGLISH).equals(this.COMPRESSION_CODEC)) {
-            fileExtension = GZIP_FILE_EXTENSION;
-            isValidCompressionEnabled = true;
+            this.fileExtension = GZIP_FILE_EXTENSION;
+            this.isValidCompressionEnabled = true;
         } else {
-            fileExtension = "";
+            this.fileExtension = "";
+            this.isValidCompressionEnabled = false;
         }
+    }
 
-        for(int chunkId = 0; chunkId < getNumChunks(); chunkId++) {
-
-            this.indexFileSizeInBytes[chunkId] = 0L;
-            this.valueFileSizeInBytes[chunkId] = 0L;
-            this.checkSumDigestIndex[chunkId] = CheckSum.getInstance(checkSumType);
-            this.checkSumDigestValue[chunkId] = CheckSum.getInstance(checkSumType);
-            this.position[chunkId] = 0;
-            this.taskIndexFileName[chunkId] = new Path(FileOutputFormat.getOutputPath(conf),
-                                                       getStoreName() + "."
-                                                               + Integer.toString(chunkId) + "_"
-                                                               + this.taskId + INDEX_FILE_EXTENSION
-                                                               + fileExtension);
-            this.taskValueFileName[chunkId] = new Path(FileOutputFormat.getOutputPath(conf),
-                                                       getStoreName() + "."
-                                                               + Integer.toString(chunkId) + "_"
-                                                               + this.taskId + DATA_FILE_EXTENSION
-                                                               + fileExtension);
-            if(this.fs == null)
-                this.fs = this.taskIndexFileName[chunkId].getFileSystem(conf);
-            if(isValidCompressionEnabled) {
-                this.indexFileStream[chunkId] = new DataOutputStream(new BufferedOutputStream(new GZIPOutputStream(fs.create(this.taskIndexFileName[chunkId]),
-                                                                                                                   DEFAULT_BUFFER_SIZE)));
-                this.valueFileStream[chunkId] = new DataOutputStream(new BufferedOutputStream(new GZIPOutputStream(fs.create(this.taskValueFileName[chunkId]),
-                                                                                                                   DEFAULT_BUFFER_SIZE)));
+    /**
+     * The MapReduce framework should operate sequentially, so thread safety shouldn't be a problem.
+     */
+    @NotThreadsafe
+    private void initFileStreams(int chunkId) {
+        /**
+         * {@link Set#add(Object)} returns false if the element already existed in the set.
+         * This ensures we initialize the resources for each chunk only once.
+         */
+        if (chunksHandled.add(chunkId)) {
```

Comment:
I believe this can introduce regression. Let us say that you have chunks 1 to 5 and when we partition the key space we only got 1,2,3 and 5 ( 4 goes missing). The way our RO code is initialized it will read 1,2,3 and will never read 5. 

Though I like what you are doing here, I would suggest keeping the regression fix to be barely minimal. We will introduce refactoring, immediately after that. Valuable lesson I learnt when we work on hot fixes before the final build.